### PR TITLE
fix false positive E0601 and E0602 with nonlocal keyword.

### PR DESCRIPTION
**Pull request by sushobhit27**
fix false positive ``undefined-variable`` and ``used-before-assignment`` with nonlocal keyword usage.

### Fixes / new features
- #2049

### CODE REVIEW

**Comment by sushobhit27** [comment]
@PCManticore any idea why this error (https://travis-ci.org/PyCQA/pylint/jobs/413143962#L730) is coming only on pypy version, this error is coming for past few days only.

**Comment by PCManticore** [comment]
@sushobhit27 Not sure, but it seems that PyPy wants to install `typed_ast` although it's marked explicitly to be skipped using environment markers. Maybe a `pip/setuptools` update that broke the environment marker support?

**Comment by brycepg** [comment]
I restarted the pypy build and the problem fixed itself?

**Comment by PCManticore** [comment]
It works now because I released astroid 2.0.3 last night. For some reason 2.0.2 didn't have environment markers for typed_ast.

File: pylint/checkers/utils.py [hunk 1]
```diff
@@ -203,6 +203,48 @@ def is_builtin(name):
     """
     return name in builtins or name in SPECIAL_BUILTINS
 
+
+def is_defined_in_scope(var_node, varname, scope):
```
Comment (PCManticore) [inline pylint/checkers/utils.py]: @sushobhit27 I think this refactoring needs to be done in a separate commit. Now it's not clear where the fix actually is, while if the function would have been moved in a separate commit, it would have been more obvious.

Comment (sushobhit27) [reply pylint/checkers/utils.py]: @PCManticore actually fix is due to below code, rest is refactoring:
```
 if isinstance(scope, astroid.If):
        for node in scope.body:
            if ((isinstance(node, astroid.AssignName) and node.name == varname)
                    or (isinstance(node, astroid.Nonlocal) and varname in node.names)):
                return True
```
So I will split it into two commits, first containing only refactor code and second one having above additional code.

File: pylint/checkers/utils.py [hunk 2]
```diff
@@ -203,6 +203,48 @@ def is_builtin(name):
...
+    if isinstance(scope, astroid.If):
+        for node in scope.body:
+            if ((isinstance(node, astroid.AssignName) and node.name == varname)
```
Comment (PCManticore) [inline pylint/checkers/utils.py]: I don't think this actually does anything, since you cannot have standalone `AssignName` nodes. Most of the time `AssignName` is wrapped in an `Assign`, so you might want to look into that instead.

File: pylint/test/functional/undefined_variable.py [hunk 3]
```diff
@@ -232,3 +232,16 @@ class LambdaClass5:
     myattr = 1
     mylambda = lambda: LambdaClass5.myattr
     mylambda4 = lambda a=LambdaClass5: lambda: a # [undefined-variable]
+
+
+def Main():
```
Comment (PCManticore) [inline pylint/test/functional/undefined_variable.py]: Let's rename to function to be clearer where this is coming from (something along the lines of `undefined_variable_issue_...` which should make it more obvious what it is testing.)

File: ChangeLog [hunk 4]
```diff
@@ -7,6 +7,9 @@ What's New in Pylint 2.2?
 
 Release date: TBA
 
+   * fix false positive ``undefined-variable`` and ``used-before-assignment`` with nonlocal keyword usage.
```
Comment (PCManticore) [inline ChangeLog]: Capitalize `fix` into `Fix`

File: doc/whatsnew/2.2.rst [hunk 5]
```diff
@@ -19,4 +19,7 @@ Other Changes
 * `unnecessary-pass` is now also emitted when a function or class contains only docstring and pass statement, 
   in which case, docstring is enough for empty definition.
 
+* fix false positive ``undefined-variable`` and ``used-before-assignment`` with nonlocal keyword usage.
```
Comment (PCManticore) [inline doc/whatsnew/2.2.rst]: Ditto. This could be:
```
Fixed a false positive for ``...`` and ``...` which happens when a `nonlocal` keyword is used and the variable is defined outside the scope where the nonlocal was used

Closes #...
```

**Comment by sushobhit27** [comment]
@PCManticore build is failing due to old-style-class check being disabled at some places in code. e.g

pylint/lib/python3.5/site-packages/pylint/checkers/spelling.py:35:0: E0012: Bad option value 'old-style-class' (bad-option-value)
I could not find this check in the master and maybe that's why when it being disabled, bad-option-value is thrown. Please let me know how to proceed further.
